internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -496,6 +510,42 @@ func (r *BMACReconciler) reconcileAgentInventory(log logrus.FieldLogger, bmh *bm
 
 }
 
+// Ask BMH to reboot the host if the agent is unbound after installation
+//
+// Here we will mark `reboot.metal3.io` annotation in the BMH, when BoundCondition
+// is equal to UnbindingPendingUserActionReason
+//
+// This will trigger a reconcile on the BMH side, resulting in node reboot
+//
+func (r *BMACReconciler) reconcileUnboundAgent(log logrus.FieldLogger, bmh *bmh_v1alpha1.BareMetalHost, agent *aiv1beta1.Agent) reconcileResult {
+	log.Debugf("Started Unbound Agent reconcile for agent %s/%s and bmh %s/%s", agent.Namespace, agent.Name, bmh.Namespace, bmh.Name)
+
+	// proceed with the reconcile only when the agent ask for user action following unbinding
+	boundCondition := conditionsv1.FindStatusCondition(agent.Status.Conditions, aiv1beta1.BoundCondition)
+	if boundCondition == nil || boundCondition.Reason != aiv1beta1.UnbindingPendingUserActionReason {
+		return reconcileComplete{}
+	}
+
+	if bmh.ObjectMeta.Annotations == nil {
+		bmh.ObjectMeta.Annotations = make(map[string]string)
+	}
+
+	// re-attach the bmh and clear the ISO url to force ironic image refresh
+	// and re-conciliation of BMH and agent
+	delete(bmh.ObjectMeta.Annotations, BMH_DETACHED_ANNOTATION)
```

Comment:
Maybe only delete if annotations is not nil?
if bmh.ObjectMeta.Annotations != nil {
		delete(bmh.ObjectMeta.Annotations, BMH_DETACHED_ANNOTATION)
}

Reply:
why? delete will not fail if it is not there and the action is idempotent

Reply:
After inspecting the code more closely: we do not need the if at all. delete will work both on nil and existing maps. Removing the redundant allocation